csrc/scheduler/matmul.cpp

```diff
@@ -51,6 +57,364 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+        "Prolog swizzle for ldmatrix, illegal tile size for prolog swizzle",
+        tile_size_x,
+        "x",
+        tile_size_y);
+
+    /* Note [How to remove bank conflict for ldmatrix?]
+     *
+     * **This note is interleaved with code, I suggest reading this note like
+     *   reading a jupyter notebook**
+     *
+     * Our task is to make sure different rows does not fall into the same
+     * bank of shared memory.
+     *
+     * Introduction to bank conflict can be found at page 54-72 of:
+     * https://on-demand.gputechconf.com/gtc/2018/presentation/s81006-volta-architecture-and-performance-optimization.pdf
+     *
+     * When we talk about bank conflict removal, we are talking about the
+     * following task:
+     *   "there are 32 banks, and each bank contains one 4-byte word, we want to
+     *    make sure different lanes in a warp does not access different word
+     *    addresses in the same bank"
+     * For example, if thread 0 is accessing word address 1, and thread 1 is
+     * accessing word address 33, then these two threads will have a bank
+     * conflict because they are accessing different word addresses in the same
+     * bank. However, if thread 0 is accessing byte address 4 and thread 1 is
+     * accessing byte address 6 then there will be no bank conflict because 4
+     * and 6 both belong to word 1.
+     */
+
+    constexpr int smem_bytes_per_word = 4;
+    constexpr int smem_banks = 32;
+
+    /* but here, for our convenience, because ldmatrix always use vectorized
+     * access of 8 items = 16 bytes = 4 words, we further group words into
+     * units: we consider each 4 words as a "unit", and each 4 banks as a
+     * "megabank". So we can rephrase our task as:
+     *   "there are 8 megabanks, and each megabanks contains one 4-word unit, we
+     *    want to make sure different lanes in a warp does not access different
+     *    unit addresses in the same megabank"
+     * In this terminology, each matrix has 8 rows, and each row has exactly one
+     * unit.
+     */
+
+    constexpr int items_per_unit = ldmatrix_cols;
+    constexpr int bytes_per_unit =
+        items_per_unit * primDataTypeSize(DataType::Half);
+    constexpr int words_per_unit = bytes_per_unit / smem_bytes_per_word;
+    constexpr int num_megabanks = smem_banks / words_per_unit;
+
+    /* In the following example, each CTA tile contains 2 rows and 3 colums of
+     * matrices, each 8x8 size:
+     *   +----------+----------+----------+
+     *   | matrix 0 | matrix 1 | matrix 2 |
+     *   +----------+----------+----------+
+     *   | matrix 3 | matrix 4 | matrix 5 |
+     *   +----------+----------+----------+
+     * The addresses of different rows in the same matrix are offseted by 3
+     * units. In this perspective, loading a matrix is a strided memory access
+     * with the following stride (in units):
+     */
+
+    int row_stride = tile_size_y / items_per_unit;
```

Comment:
or `units_per_row`. (not suggesting to change the name. just thought it'd be helpful to have this mentioned)

Reply:
added as comment